google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -181,6 +181,258 @@ class ResumableAsyncStreamingReadWriteRpc {
+    future<void> root_future = root_promise.get_future();
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      switch (stream_state_) {
+        case State::kShutdown:
+          return make_ready_future();
+        case State::kRetrying:
+          stream_state_ = State::kShutdown;
+          root_future = root_future.then(ChainFuture(
+              retry_promise_->get_future().then([this](future<void>) {
+                std::lock_guard<std::mutex> g{mu_};
+                CompleteStreamLockHeld(Status());
+              })));
+          break;
+        case State::kInitialized:
+          stream_state_ = State::kShutdown;
+          if (in_progress_read_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_read_->get_future()));
+          }
+          if (in_progress_write_) {
+            root_future =
+                root_future.then(ChainFuture(in_progress_write_->get_future()));
+          }
+          std::shared_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>
+              stream = std::move(stream_);
+          CompleteStreamLockHeld(Status());
+          root_future = root_future.then([stream](future<void>) {
+            return future<void>(stream->Finish());
+          });
+      }
+    }
+    root_promise.set_value();
+    return root_future;
+  }
+
+ private:
+  using UnderlyingStream =
+      std::unique_ptr<AsyncStreamingReadWriteRpc<RequestType, ResponseType>>;
+
+  enum class State { kRetrying, kInitialized, kShutdown };
+
+  void SetReadWriteFuturesLockHeld() {
+    absl::optional<promise<void>> read_reinit_done;
+    absl::optional<promise<void>> write_reinit_done;
+
+    {
+      read_reinit_done.swap(read_reinit_done_);
+      write_reinit_done.swap(write_reinit_done_);
+    }
+    mu_.unlock();
+    if (read_reinit_done) read_reinit_done->set_value();
+
+    if (write_reinit_done) write_reinit_done->set_value();
+    mu_.lock();
+  }
+
+  void CompleteStreamLockHeld(Status status) {
+    SetReadWriteFuturesLockHeld();
+    status_promise_.set_value(std::move(status));
+  }
+
+  void FinishRetryPromiseLockHeld() {
+    assert(retry_promise_.has_value());
+    promise<void> retry_promise = std::move(*retry_promise_);
+    retry_promise_.reset();
+    mu_.unlock();
+    retry_promise.set_value();
+    mu_.lock();
+  }
+
+  void AttemptRetry(Status const& status,
+                    std::shared_ptr<RetryPolicy> retry_policy,
+                    std::shared_ptr<BackoffPolicy> backoff_policy) {
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (stream_state_ == State::kShutdown) {
+        return FinishRetryPromiseLockHeld();
+      }
+      assert(stream_state_ == State::kRetrying);
+    }
+
+    if (!retry_policy->IsExhausted() && retry_policy->OnFailure(status)) {
+      sleeper_(backoff_policy->OnCompletion())
+          .then([this, retry_policy, backoff_policy](future<void>) {
+            Initialize(retry_policy, backoff_policy);
```

Comment:
`sleeper_` is just supposed to invoke a pure time delay.

Reply:
That won't work.  You would be blocking the thread, i.e., the thread cannot do any other useful work, while sleeping. This can be a problem because typically there is a limited number of background threads dedicated to this work.

You should use this:

 https://github.com/googleapis/google-cloud-cpp/blob/291e9dd4d3a66ecb3133d79ae58abff3491079cc/google/cloud/completion_queue.h#L95-L97

Reply:
This is going to invoke MakeRelativeTimer under the hood, but is a smaller, simpler API. There's no reason to care if the completion queue is shut down and this returns an error- that just means we'll create a new stream, which should fail with a permanent error if, for example, the completion queue is shut down.

I don't think there's any need to handle more complexity here.

Reply:
Looking at my own code, I think you are right.  For some reason I recall that not working in the past.